[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor.

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66604
* #66603

This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

Commit:
[PyTorch][1/N] Basic implementation of ShardedEmbedding and ShardedEmbeddingBag using ShardedTensor. 

This diff/PR is trying to implement the ShardedEmbedding and ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.
4. This is just col-wise sharding, row-wise sharding is on the way.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding and ShardedEmbeddingBag using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding and ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.
4. This is just col-wise sharding, row-wise sharding is on the way.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding and ShardedEmbeddingBag using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding and ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.
4. This is just col-wise sharding, row-wise sharding is on the way.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding and ShardedEmbeddingBag using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding and ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.
4. This is just col-wise sharding, row-wise sharding is on the way.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 reviewed (commented):
Awesome work! Just a few comments inline.

bowangbj reviewed (commented):
Thanks Junjie, great progress! 
One interesting thing (can be a follow up PR) is to figure out how this PR can be fit together with D31459953, probably D31459953 can be built on top of this PR

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding and ShardedEmbeddingBag using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding and ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.
4. This is just col-wise sharding, row-wise sharding is on the way.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

Commit:
Update on "[PyTorch][1/N] Basic implementation of ShardedEmbedding using ShardedTensor. "


This diff/PR is trying to implement the ShardedEmbedding using the ShardedTensor.

Several caveats:
1. We support limited input params for these two ops. To support more params are on the way.
2. We only support chuck sharding for now.
3. We only support a single local shard per rank for now.

Differential Revision: [D31544556](https://our.internmc.facebook.com/intern/diff/D31544556/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31544556/)!

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]
